neqo-transport/src/server.rs

```diff
@@ -620,71 +561,68 @@ impl Server {
+                Output::None => {}
+                d @ Output::Datagram(_) => return d,
+                Output::Callback(next) => match callback {
+                    Some(previous) => callback = Some(min(previous, next)),
+                    None => callback = Some(next),
+                },
             }
         }
 
-        qtrace!([self], "No packet to send still, check wake up times");
-        loop {
-            let connection = self
-                .connections
-                .borrow()
-                .values()
-                .find(|c| c.borrow().needs_waking(now))
-                .cloned()?;
-            let datagram = self.process_connection(&connection, None, now);
-            connection.borrow_mut().woken();
-            if datagram.is_some() {
-                return datagram;
-            }
-        }
+        callback.map_or(Output::None, Output::Callback)
     }
 
     pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
-        if self.wake_at.map_or(false, |c| c <= now) {
-            self.wake_at = None;
-        }
+        let out = dgram
+            .map_or(Output::None, |d| self.process_input(d, now))
+            .or_else(|| self.process_next_output(now));
+
+        // Clean-up closed connections.
+        self.connections.borrow_mut().retain(|_, c| {
+            if matches!(c.borrow().state(), State::Closed(_)) {
+                // TODO: Is this still needed?
+                c.borrow_mut().set_qlog(NeqoQlog::disabled());
+                false
+            } else {
+                true
+            }
+        });
 
-        dgram
-            .and_then(|d| self.process_input(d, now))
-            .or_else(|| self.process_next_output(now))
-            .map(|d| {
-                qtrace!([self], "Send packet: {:?}", d);
-                Output::Datagram(d)
-            })
-            .or_else(|| self.wake_at.take().map(|c| Output::Callback(c - now)))
-            .unwrap_or_else(|| {
-                qtrace!([self], "Go dormant");
-                Output::None
-            })
+        out
     }
 
     /// This lists the connections that have received new events
     /// as a result of calling `process()`.
-    pub fn active_connections(&mut self) -> Vec<ActiveConnectionRef> {
-        mem::take(&mut self.active).into_iter().collect()
+    pub fn active_connections(&mut self) -> HashSet<ActiveConnectionRef> {
+        self.connections
+            .borrow()
+            .values()
+            .filter(|c| c.borrow().has_events())
+            .map(|c| ActiveConnectionRef { c: Rc::clone(c) })
+            .collect()
     }
 
     /// Whether any connections have received new events as a result of calling
     /// `process()`.
     #[must_use]
     pub fn has_active_connections(&self) -> bool {
-        !self.active.is_empty()
-    }
-
-    pub fn add_to_waiting(&mut self, c: &ActiveConnectionRef) {
-        self.waiting.push_back(c.connection());
```

Comment:
Given that now `Server` always `process`es each `Connection`, no need for users to tell `Server` that a connection might be waiting for processing.